[BasicBlockSections] Apply path cloning with -basic-block-sections.

rlavaee opened the pull request:
https://github.com/llvm/llvm-project/commit/28b912687900bc0a67cd61c374fce296b09963c4 introduced the path cloning format in the basic-block-sections profile.

This PR validates and applies path clonings. 
A path cloning is valid if all of these conditions hold:
  1. All bb ids in the path are mapped to existing blocks.
  2. Each two consecutive bb ids in the path have a successor relationship in the CFG.
  3. The path does not include a block with indirect branches, except possibly as the last block.
 
Applying a path cloning involves cloning all blocks in the path (except the first one) and setting up their branches.
Once all clonings are applied, the cluster information is used to guide block layout in the modified function.

rlavaee committed:
[BasicBlockSections] Apply path cloning with -basic-block-sections.

rlavaee committed:
Apply BB sections even if cloning fails.

spupyrev commented:
Do you have some recent perf numbers for this kind of optimization? We've tried something similar internally in the past but didn't see any perf wins; i wonder if it makes sense for us to reconsider the optimization.

rlavaee commented:
> Do you have some recent perf numbers for this kind of optimization? We've tried something similar internally in the past but didn't see any perf wins; i wonder if it makes sense for us to reconsider the optimization.

We got 2% improvement on clang (over Propeller-optimized). For internal benchmarks, we got 0.4% improvement on Search before, but I will need to refresh that result. The analysis part of this is not released externally though.

tmsri commented on llvm/lib/CodeGen/BasicBlockSections.cpp:
```diff
@@ -99,6 +101,33 @@ static cl::opt<bool> BBSectionsDetectSourceDrift(
+  for (auto &I : MBB->instrs())
+    CloneBB->push_back(MF.CloneMachineInstr(&I));
+
+  // Add the successors of the original block as the new block's successors.
```
Why not predecessors here?

rlavaee replied:
The single predecessor for this block is set after returning from this call.

tmsri replied:
Sure, sound good.

tmsri commented on llvm/lib/CodeGen/BasicBlockSections.cpp:
```diff
@@ -99,6 +101,33 @@ static cl::opt<bool> BBSectionsDetectSourceDrift(
 
 namespace {
 
+MachineBasicBlock *CloneMachineBasicBlock(MachineBasicBlock *MBB) {
```
Why can't this be part of the IR itself, why not in BasicBlock.h?

rlavaee replied:
We need to copy the machine-instructions.

tmsri replied:
I mean CodeGen/MachineBasicBlock.h.   This is a simple change so I am not insisting.

rlavaee replied:
I see. MachineFunction.h is the proper place. I augmented `MachineFunction::createMachineBasicBlock` so we can piggy-back on it.

tmsri commented on llvm/lib/CodeGen/BasicBlockSections.cpp:
```diff
@@ -285,6 +314,130 @@ static bool hasInstrProfHashMismatch(MachineFunction &MF) {
+// from `ProfileBBID`s of all clone blocks to their BBIDs (assigned by
+// `MachineFunction`).
+static DenseMap<ProfileBBID, unsigned>
+ApplyCloning(MachineFunction &MF,
```
Why should cloning be clubbed with BasicBlock Sections pass?  I am not sure this is the right place for it, maybe cloning can be made a separate pass?

rlavaee replied:
Great idea. I moved this to a new pass and augmented `MachineBasicBlock::BBID` to have a `unsigned CloneID` so we can correctly map in the BasicBlockSections pass.

tmsri commented on llvm/lib/CodeGen/BasicBlockSections.cpp:
```diff
@@ -285,6 +314,130 @@ static bool hasInstrProfHashMismatch(MachineFunction &MF) {
 
+// Returns if we can legally apply all clonings specified in `ClonePaths`.
+static bool
+IsValidCloning(const MachineFunction &MF,
```
Please describe the parameters

tmsri commented on llvm/lib/CodeGen/BasicBlockSections.cpp:
```diff
@@ -285,6 +314,130 @@ static bool hasInstrProfHashMismatch(MachineFunction &MF) {
+      WithColor::warning()
+          << "block #" << BBID
+          << " has indirect branch and appears as the non-tail block of a "
+             "path in function "
```
What are you trying to prevent here?

rlavaee replied:
We cannot clone a block that is in the middle of a path and has an indirect branch.

tmsri commented on llvm/lib/CodeGen/BasicBlockSections.cpp:
```diff
@@ -285,6 +314,130 @@ static bool hasInstrProfHashMismatch(MachineFunction &MF) {
+    for (unsigned BBID : ClonePath) {
+      MachineBasicBlock *OrigBB = BBIDToBlock.at(BBID);
+      if (PrevBB == nullptr) {
+        if (auto FT = OrigBB->getFallThrough(/*JumpToFallThrough=*/false)) {
```
Why is this done only for the first block in the clone path?

tmsri commented on llvm/lib/CodeGen/BasicBlockSections.cpp:
```diff
@@ -285,6 +314,130 @@ static bool hasInstrProfHashMismatch(MachineFunction &MF) {
+      CloneBBIDMap.try_emplace({BBID, ++NClonesForBBID[BBID]},
+                               *CloneBB->getBBID());
+      // Set up the previous block in the path to jump to the clone.
+      PrevBB->ReplaceUsesOfBlockWith(OrigBB, CloneBB);
```
Ok so this is where the predecessors get updated too, please ignore the original comment then.

tmsri commented on llvm/test/CodeGen/X86/basic-block-sections-cloning.ll:
```diff
@@ -0,0 +1,185 @@
+; RUN: llc < %s -mtriple=x86_64-pc-linux -O0 -function-sections -basic-block-sections=%t4 2> %t4.err | FileCheck %s --check-prefixes=FOONOCLONE,FOOSECTIONS
+; RUN: FileCheck %s --check-prefixes=PATH4-WARN < %t4.err
+
+define void @foo(i1 %a, i1 %b, i1 %c, i1 %d) {
```
Can you add a comment on which call instruction corresponds to which bbid?

rlavaee replied:
The test will check the name of the blocks (emitted as comment after the block's label). Please let me know if you have other suggestions.

rlavaee committed:
Introduce a new pass for path cloning.

rlavaee committed:
clang-format.

rlavaee committed:
Add clone id to MIR serialization.

rlavaee committed:
Address remaining comments.

rebase.

rlavaee committed:
Fix the mbb profile dump test.

MatzeB commented on llvm/lib/CodeGen/BasicBlockPathCloning.cpp:
```diff
@@ -0,0 +1,227 @@
+//
+//===----------------------------------------------------------------------===//
+//
+// BasicBlockPathCloning implementation.
```
Use 
```
/// \file
/// ...
```
to expose the commen to doxygen.

MatzeB commented on llvm/include/llvm/CodeGen/BasicBlockSectionUtils.h:
```diff
@@ -27,6 +27,13 @@ void sortBasicBlocksAndUpdateBranches(MachineFunction &MF,
 
 void avoidZeroOffsetLandingPad(MachineFunction &MF);
 
+// This checks if the source of this function has drifted since this binary was
```
Use `///` to expose comment to doxygen.

MatzeB commented on llvm/lib/CodeGen/BasicBlockPathCloning.cpp:
```diff
@@ -0,0 +1,227 @@
+
+  // Copy the instructions.
+  for (auto &I : OrigBB.instrs())
+    CloneBB->push_back(MF.CloneMachineInstr(&I));
```
I see a `TargetInstrInfo::duplicate` API that appears to add functionality on top of `MachineFunction::CloneMachineInstr`?

rlavaee replied:
It looks like this only clones bundled instructions (https://llvm.org/doxygen/MachineFunction_8cpp_source.html#l00418), while I want all instructions cloned.

MatzeB replied:
I think you should rather read this as it cannot clone instructions within a bundle, but apart from that you can probably consider every instruction as a bundle of one for this API.

MatzeB replied:
The only Target with an interesting implementation (the default version just calls `CloneMachineInstr` anyway), seems to be ARM here: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/ARM/ARMBaseInstrInfo.cpp#L1836 which seems to duplicate constant pool values so they can be kept close to the instruction and avoid the risk of ARM immediates being too small? So probably worth using the API?

rlavaee replied:
I see. You meant using the `duplicate` API instead of  `CloneMachineInstr`. That makes sense. I adopted the API but had to relax an assertion in its implementation to allow copying CFI instructions. (CFI instructions are marked non-duplicable because of Darwin: https://github.com/llvm/llvm-project/blob/49504674dbca6f753771f790ea4062bcbc9cc829/llvm/lib/CodeGen/TailDuplicator.cpp#L608). I also added another validation for nonduplicable instructions.

MatzeB commented on llvm/lib/CodeGen/BasicBlockPathCloning.cpp:
```diff
@@ -0,0 +1,227 @@
+  for (size_t I = 0; I < ClonePath.size(); ++I) {
+    unsigned BBID = ClonePath[I];
+    const MachineBasicBlock *PathBB = BBIDToBlock.lookup(BBID);
+    if (!PathBB) {
+      WithColor::warning() << "no block with id " << BBID << " in function "
+                           << MF.getName() << "\n";
+      return false;
+    }
+
+    if (PrevBB && !PrevBB->isSuccessor(PathBB)) {
+      WithColor::warning() << "block #" << BBID
+                           << " is not a successor of block #"
+                           << PrevBB->getBBID()->BaseID << " in function "
+                           << MF.getName() << "\n";
+      return false;
+    }
+
+    if (I != ClonePath.size() - 1 && !PathBB->empty() &&
+        PathBB->back().isIndirectBranch()) {
+      WithColor::warning()
+          << "block #" << BBID
+          << " has indirect branch and appears as the non-tail block of a "
+             "path in function "
+          << MF.getName() << "\n";
+      return false;
```
I find it unusual to see "warnings" in the backend. We tend to go for a fail fast coding style with `assert`s or `report_fatal_error`s usually and don't try to continue when there is invalid conditions...

rlavaee replied:
This is more like a PGO hash mismatch warning.  We proceed with applying other clonings even though one may fail. Do you suggest I use `dbgs()` instead?

MatzeB reviewed (commented):
did not end up doing a deep review. But added some nits :)

rlavaee committed:
Address comments.

rlavaee reviewed (commented):
Thanks for the review @MatzeB

rlavaee committed:
Use TargetInstrInfo::duplicate to clone instructions.

Also check for non-duplicable instructions in the block similar to
TailDuplicator.cpp.

rlavaee committed:
Check to avoid double-copying bundled instrs.

tmsri commented on llvm/test/CodeGen/X86/basic-block-sections-cloning.ll:
```diff
@@ -7,35 +7,35 @@ declare void @effect(i32 zeroext)
 ; RUN: echo 'c 0 3.1 5.1 1 2 3 4 5' >> %t1
-; RUN: llc < %s -mtriple=x86_64-pc-linux -O0 -function-sections -basic-block-sections=%t1 | FileCheck %s --check-prefixes=PATH1,FOOSECTIONS,FOOCLONE
+; RUN: llc < %s -mtriple=x86_64-pc-linux -O0 -function-sections -enable-basic-block-path-cloning -basic-block-sections=%t1 | FileCheck %s --check-prefixes=PATH1,FOOSECTIONS,FOOCLONE
 ; RUN: echo 'v1' > %t2
```
Please split this into multiple tests in different files so that it is easier to maintain going forward.

rlavaee replied:
Thank you. Done.

rlavaee committed:
Split tests into four files.

rlavaee committed:
Remove the unnecessary -mllvm flag.

rlavaee committed:
Copy the livein set.

rlavaee committed:
Split one more test into separate files.

rlavaee commented on llvm/test/CodeGen/X86/basic-block-sections-cloning.ll:
```diff
@@ -0,0 +1,111 @@
+; RUN: echo 'c 0 3.1 5.1 1 2 3 4 5' >> %t1
+; RUN: llc < %s -mtriple=x86_64-pc-linux -O0 -function-sections -enable-basic-block-path-cloning -basic-block-sections=%t1 | FileCheck %s --check-prefixes=PATH1
+; RUN: llc < %s -mtriple=x86_64-pc-linux -O0 -function-sections -enable-basic-block-path-cloning -basic-block-sections=%t1 -stop-after=bb-path-cloning | FileCheck %s --check-prefix=PATH1MIR
+; RUN: echo 'v1' > %t2
```
Done. Thanks for the review.

spupyrev commented:
@rlavaee could you please point to the code where these cloned paths are being implemented? There is likely a non-trivial tradeoff between eliminating branches/jumps and increasing the code size. I'm curious to see the algorithm

rlavaee commented:
> @rlavaee could you please point to the code where these cloned paths are being implemented? There is likely a non-trivial tradeoff between eliminating branches/jumps and increasing the code size. I'm curious to see the algorithm

Thanks for the interest. We will open-source the implementation under https://github.com/google/autofdo in a couple months.
